### PR TITLE
Add support for STM32F767xx.

### PR DESCRIPTION
**Pull request by bartek-w**
Built-in support for STM32F767 tested with NUCLEO-F767ZI board.

When connecting in low power modes, internal bus is likely not allowing to access memory/device/core.
It looks like recommended connection mode for STM32 doesn't fit well to existing pyOCD architecture (please suggest simpler solution without MiniAP if possible).
Most of the code probably should be generalized to all STM32F family.
Code successfully allows to connect with halt to chip running low power mode code (also after first connection, it is possible to attach to core without special reset until power cycle).

WITHOUT the specialised code to connect under reset, set debug mode and halt core, you can see multiple different errors:
(many of these errors are mentioned on open issues related to stm32).
Same error codes without halt before discovery includes wrong cidr (basically read of the value can be random, as there is no clock for AHP).
```

0001459:DEBUG:sequencer:Running task create_ap.0
0001461:DEBUG:ap:Using accelerated memory access interface
0001461:INFO:ap:AHB-AP#0 IDR = 0x74770001 (AHB-AP var0 rev7)
0001463:DEBUG:ap:AHB-AP#0 default HPROT=b HNONSEC=0
0001465:DEBUG:ap:AHB-AP#0 implemented HPROT=f HNONSEC=1
0001469:DEBUG:sequencer:Running task find_components
0001469:DEBUG:sequencer:Running task init_ap.0
0001474:WARNING:rom_table:Invalid coresight component, cidr=0x0 <- this can be almost random value
0001475:DEBUG:sequencer:Running task create_cores
0001475:DEBUG:sequencer:Running task create_components
0001475:DEBUG:sequencer:Running task check_for_cores
0001475:DEBUG:session:uninit session <pyocd.core.session.Session object at 0x10c539890>
0001478:CRITICAL:__main__:No cores were discovered!

```
Another error type is complete lack of power:
```

0001878:DEBUG:sequencer:Running task connect
0001880:DEBUG:session:uninit session <pyocd.core.session.Session object at 0x10daa6d90>
0001882:CRITICAL:__main__:STLink error (11): Debug power error
Traceback (most recent call last):
  File "//pyocd/__main__.py", line 368, in run
    self._COMMANDS[self._args.cmd](self)
  File "//pyocd/__main__.py", line 525, in do_flash
    with session:
  File "//pyocd/core/session.py", line 302, in __enter__
    self.open()
  File "//pyocd/core/session.py", line 420, in open
    self._board.init()
  File "//pyocd/board/board.py", line 83, in init
    self.target.init()
  File "//pyocd/core/coresight_target.py", line 160, in init
    seq.invoke()
  File "//pyocd/utility/sequencer.py", line 213, in invoke
    resultSequence.invoke()
  File "//pyocd/utility/sequencer.py", line 208, in invoke
    resultSequence = call()
  File "//pyocd/coresight/dap.py", line 278, in _connect
    connector.connect(self._protocol)
  File "//pyocd/coresight/dap.py", line 147, in connect
    self._probe.connect(protocol)
  File "//pyocd/probe/stlink_probe.py", line 126, in connect
    self._link.enter_debug(STLink.Protocol.SWD)
  File "//pyocd/probe/stlink/stlink.py", line 256, in enter_debug
    self._check_status(response)
  File "//pyocd/probe/stlink/stlink.py", line 294, in _check_status
    raise exceptions.ProbeError(error_message)
pyocd.core.exceptions.ProbeError: STLink error (11): Debug power error
```

It is also not possible to run connect_mode=under_reset for STM32F7:

```
0001496:DEBUG:ap:Using accelerated memory access interface
0001496:INFO:ap:AHB-AP#0 IDR = 0x74770001 (AHB-AP var0 rev7)
0001499:DEBUG:ap:AHB-AP#0 default HPROT=b HNONSEC=0
0001500:DEBUG:ap:AHB-AP#0 implemented HPROT=f HNONSEC=1
0001503:DEBUG:sequencer:Running task find_components
0001504:DEBUG:sequencer:Running task init_ap.0
0001509:INFO:rom_table:AHB-AP#0 Class 0x1 ROM table #0 @ 0xe00fd000 (designer=020 part=451)
0001513:INFO:rom_table:[0]<e00fe000:ROM class=1 designer=43b part=4c8>
0001513:INFO:rom_table:  AHB-AP#0 Class 0x1 ROM table #1 @ 0xe00fe000 (designer=43b part=4c8)
0001518:INFO:rom_table:  [0]<e00ff000:ROM class=1 designer=43b part=4c7>
0001518:INFO:rom_table:    AHB-AP#0 Class 0x1 ROM table #2 @ 0xe00ff000 (designer=43b part=4c7)
0001523:ERROR:rom_table:Error attempting to probe CoreSight component referenced by ROM table entry #0: SWD/JTAG Transfer Fault @ 0xe000efbc-0xe000f043
Traceback (most recent call last):
  File "//pyocd/coresight/rom_table.py", line 349, in _read_table
    self._handle_table_entry(entry, entryNumber)
  File "//pyocd/coresight/rom_table.py", line 399, in _handle_table_entry
    cmpid.read_id_registers()
  File "//pyocd/coresight/rom_table.py", line 111, in read_id_registers
    regs = self.ap.read_memory_block32(self.top_address + self.IDR_READ_START, self.IDR_READ_COUNT)
  File "//pyocd/probe/stlink_probe.py", line 260, in read_memory_block32
    return conversion.byte_list_to_u32le_list(self._link.read_mem32(addr, size * 4, self._apsel))
  File "//pyocd/probe/stlink/stlink.py", line 374, in read_mem32
    return self._read_mem(addr, size, Commands.JTAG_READMEM_32BIT, self.MAXIMUM_TRANSFER_SIZE, apsel)
  File "//pyocd/probe/stlink/stlink.py", line 332, in _read_mem
    raise exc
pyocd.core.exceptions.TransferFaultError: SWD/JTAG Transfer Fault @ 0xe000efbc-0xe000f043
0001528:INFO:rom_table:    [1]<e0001000:DWT class=14 designer=43b part=002>
0001531:INFO:rom_table:    [2]<e0002000:FPB class=14 designer=43b part=00e>
0001533:INFO:rom_table:    [3]<e0000000:ITM class=14 designer=43b part=001>
0001536:INFO:rom_table:  [1]<e0041000:ETM-M7 class=9 designer=43b part=975 devtype=13 archid=4a13 devid=0:0:0>
0001538:INFO:rom_table:[1]<e0040000:TPIU-M7 class=9 designer=43b part=9a9 devtype=11 archid=0000 devid=ca1:0:0>
0001538:DEBUG:sequencer:Running task create_cores
0001539:DEBUG:sequencer:Running task create_components
0001539:DEBUG:discovery:Creating DWT component
0001541:INFO:dwt:4 hardware watchpoints
0001547:DEBUG:discovery:Creating FPB component
0001548:INFO:fpb:8 hardware breakpoints, 1 literal comparators
0001549:DEBUG:fpb:fpb has been disabled
0001560:DEBUG:discovery:Creating ITM component
0001566:DEBUG:discovery:Creating TPIU-M7 component
0001567:DEBUG:sequencer:Running task check_for_cores
Exception while initing board: No cores were discovered!
```

### CODE REVIEW

**Comment by flit** [comment]
Thanks again! 😄 Seems like a good solution to me. It would be nice to standardize this MiniAP architecture so it's easily reusable in other targets and situations.

I'm curious why the F767 faults when trying to read CoreSight IDs for the M7 core SCS component (that the TransferFaultError above). That might be due to a power domain that's not enabled. Notably, most other STM32 devices work fine with connect under reset.

It's also interesting to note that the CMSIS-Pack for STM32F7 does not include any special debug logic like this for recovering from low power. So I wonder if Keil MDK has similar issues (I'm following up internally).

Fwiw, the CoreSight architecture supports ROM table entries specifying a power domain that must be enabled in order to access the referenced component, and there is a Granular Power Requestor component used for this purpose (and in ADIv6, the ROM table itself optionally has power domain controls). This is supported in pyOCD, but obviously not very useful if not included. (One of my longer term missions at Arm is to help silicon vendors improve debug over low power features in their MCUs… got a long way to go… 😞 )

File: pyocd/target/builtin/target_STM32F767xx.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
+# pyOCD debugger
+# SPDX-License-Identifier: Apache-2.0
```
Comment (flit) [inline pyocd/target/builtin/target_STM32F767xx.py:2]: Can you please add your copyright here? (Technically should be added to every file you touch. But a file needs at least one copyright statement.)

File: pyocd/target/builtin/target_STM32F767xx.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    def post_connect_hook(self):
+        FLASH_OPTCR_ADDR = 0x40023C14
+        FLASH_OPTCR_NDBANK = 1<<29
+        flash_optcr = self.read32(FLASH_OPTCR_ADDR)
+        if flash_optcr & FLASH_OPTCR_NDBANK:
+            LOG.info('Single bank configuration detected [FLASH_OPTCR=0x%08x].', flash_optcr)
+        else:
+            LOG.error('Dual bank configuration detected [FLASH_OPTCR=0x%08x]. '
+                      'Currently only single bank configrautin is fully supported!', flash_optcr)
```
Comment (flit) [inline pyocd/target/builtin/target_STM32F767xx.py:187]: Minor: misspelled "configrautin".

File: pyocd/target/builtin/target_STM32F767xx.py (hunk 1 above)
Comment (bartek-w) [reply pyocd/target/builtin/target_STM32F767xx.py:2]: I'm not good at legal things: Who should be copyright holder?
I don't see any private copyrights notices in source code... does it mean that there is no private contributors to pyOCD ?

File: pyocd/target/builtin/target_STM32F767xx.py (hunk 2 above)
Comment (bartek-w) [reply pyocd/target/builtin/target_STM32F767xx.py]: Done.

**Comment by bartek-w** [comment]
> It's also interesting to note that the CMSIS-Pack for STM32F7 does not include any special debug logic like this for recovering from low power. So I wonder if Keil MDK has similar issues (I'm following up internally).

Maybe I miss something, but I wasn't able to communicate with the chip without hardware reset. It would be interesting to know if any other software/hardware can connect to STM32F7 without reset, when chip is on low power mode and DBGMCU wasn't set previously.

> I'm curious why the F767 faults when trying to read CoreSight IDs for the M7 core SCS component (that the TransferFaultError above). That might be due to a power domain that's not enabled. Notably, most other STM32 devices work fine with connect under reset.

> Fwiw, the CoreSight architecture supports ROM table entries specifying a power domain that must be enabled in order to access the referenced component, and there is a Granular Power Requestor component used for this purpose (and in ADIv6, the ROM table itself optionally has power domain controls). This is supported in pyOCD, but obviously not very useful if not included. (One of my longer term missions at Arm is to help silicon vendors improve debug over low power features in their MCUs… got a long way to go… 😞 )

I can dig into it to understand the problem better latter. It looks like I was unlucky picking the board for experiments if this issue is not common across STM32F series.

If it is working correctly, what about adding a test for connect under a reset? Also travis-ci looks like internal server... is it possible to see/track results externaly ?

File: pyocd/target/builtin/target_STM32F767xx.py (hunk 1 above)
Comment (flit) [reply pyocd/target/builtin/target_STM32F767xx.py:2]: Depends. If you're doing this on your own time, use your name as the copyright holder. If it's on employer time, use the company name.

You're right, there aren't private copyrights, but that's mostly a failure on the reviewers' part. I'm trying to improve on it. There have been plenty of private contributions, but in the past the contributor rules were different and I think work was assigned to Arm. Even my own contributions from before I joined Arm were assigned to Arm.

Comment (bartek-w) [reply pyocd/target/builtin/target_STM32F767xx.py:3]: It's a private contribution - professionally I'm ML engineer, but sometimes I miss low level programming ;)

Thanks for making the project more friendly to open source community!

**Comment by bartek-w** [comment]
Two small updates:

-  STM32CubeProgrammer also can't attach on running to sleeping target 

- 0xe000exxx just can't be accessed with reset asserted - it doesn't look like it is related to power (or at least it is non trivial). I don't have idea to debug it more - maybe this is something in stm32f7 chip design. As 'halt' (with implemented reset) connect mode provides successful connection under every condition I will not debug it more, unless you have some quick ideas.

**Comment by bartek-w** [comment]
Extra change (sorry for late addition): under-reset will also do this trick and halt core early before discovery, so Mbed Studio will be happy in default configuration.

**Comment by flit** [comment]
Thanks for trying out STM32CubeProgrammer. This behaviour seems specific to the F7 (and maybe H7) series.

And thanks for the under-reset change, I was going to suggest that.

Unfortunately the ci-morph Jenkins server is not publicly accessible, as you've discovered. I'd like to change to something like Azure Pipelines where the test results can be publicly viewed, but it's just a question of having time.

**Pull request merged**